shared/bazel/rules/cc_rules.bzl

```diff
@@ -31,30 +44,298 @@ def third_party_cc_lib_helper(
+        strip_include_prefix = strip_include_prefix,
+        **kwargs
+    )
+
     cc_library(
         name = name,
+        hdrs = hdrs,
+        copts = copts,
         srcs = srcs + [lib + "-srcs" for lib in third_party_libraries],
         deps = deps + [lib + "-headers" for lib in third_party_libraries + third_party_header_only_libraries],
+        strip_include_prefix = strip_include_prefix,
+        **kwargs
+    )
+
+    if srcs_pkg_root:
+        pkg_files(
+            name = name + "-srcs-pkg",
+            srcs = native.glob([srcs_pkg_root + "/**"]),
+            strip_prefix = srcs_pkg_root,
+        )
+
+        pkg_zip(
+            name = name + "-srcs-zip",
+            srcs = maybe_license_pkg + extra_src_pkg_files + [name + "-srcs-pkg"] + [lib + "-srcs-pkg" for lib in third_party_libraries],
+            tags = ["no-remote"],
+        )
+
+    if hdrs_pkg_root:
+        pkg_files(
+            name = name + "-hdrs-pkg",
+            srcs = native.glob([hdrs_pkg_root + "/**"]),
+            strip_prefix = hdrs_pkg_root,
+        )
+
+        pkg_zip(
+            name = name + "-hdrs-zip",
+            srcs = extra_hdr_pkg_files + maybe_license_pkg + [name + "-hdrs-pkg"] + [lib + "-hdrs-pkg" for lib in third_party_libraries + third_party_header_only_libraries],
+            tags = ["no-remote"],
+        )
+
+CcStaticLibraryInfo = provider(
+    "Information about a cc static library.",
+    fields = {
+        "linker_input": "the resulting linker input artifact for the static library",
+        "used_objects": "the object files already accounted for",
+    },
+)
+
+def _accumulate_used_objects(ctx):
+    transitive_used_objects = []
+    for dep in ctx.attr.static_deps:
+        transitive_used_objects.append(dep[CcStaticLibraryInfo].used_objects)
+
+    return transitive_used_objects
+
+def _filter_inputs(
+        ctx,
+        feature_configuration,
+        cc_toolchain,
+        deps,
+        used_objects):
+    dependency_linker_inputs_sets = []
+    for dep in deps:
+        dependency_linker_inputs_sets.append(dep[CcInfo].linking_context.linker_inputs)
+
+    dependency_linker_inputs = depset(transitive = dependency_linker_inputs_sets, order = "topological").to_list()
+
+    used_objects_depset = depset(transitive = used_objects, order = "topological").to_list()
+
+    linker_inputs = []
+    for linker_input in dependency_linker_inputs:
+        for lib in linker_input.libraries:
+            for o in lib.objects + lib.pic_objects:
+                if o not in used_objects_depset:
+                    linker_inputs.append(o)
+
+    return sorted(linker_inputs)
+
+def _cc_static_library_impl(ctx):
+    deps = ctx.attr.deps
+
+    cc_toolchain = find_cpp_toolchain(ctx)
+    feature_configuration = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features + ["force_no_whole_archive"],
```

Comment:
Mind a quick comment on what force_no_whole_archive does for us here?

Reply:
I copied this from your PR. I honestly have no idea